src/classes/BusinessLogicServices/ShopOrderService.php

```diff
@@ -0,0 +1,335 @@
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function getOrderAndShippingData($orderId)
+    {
+        $order = new Order();
+        try {
+            $sourceOrder = $this->getOrder($orderId);
+        } catch (OrderNotFound $e) {
+            Logger::logWarning(TranslationUtility::__('Source order not found'), 'Integration');
+
+            return $order;
+        }
+
+        $currencyId = (int)$sourceOrder->id_currency;
+        $currency = \Currency::getCurrency($currencyId);
+
+        $order->setId($orderId);
+        $order->setCustomerId((int)$sourceOrder->id_customer);
+        $order->setCurrency($currency['iso_code']);
+        $order->setTotalPrice((float)$sourceOrder->total_paid_tax_incl);
+        $order->setBasePrice((float)$sourceOrder->total_paid_tax_excl);
+
+        $dropOffId = $this->getDropOffId($sourceOrder);
+
+        if ($dropOffId) {
+            $order->setShippingDropOffId($dropOffId);
+        }
+
+        $order->setShippingAddress($this->getAddress($sourceOrder));
+
+        $this->setOrderShippingDetails($order, $sourceOrder->id_carrier);
+        $order->setItems($this->getOrderItems($sourceOrder));
+
+        return $order;
+    }
+
+    /**
+     * Retrieves drop-off id if shop order has drop off shipping service selected.
+     *
+     * Returns null otherwise.
+     *
+     * @param PrestaShopOrder $shopOrder
+     *
+     * @return string | null
+     */
+    private function getDropOffId(PrestaShopOrder $shopOrder)
+    {
+        try {
+            $repository = RepositoryRegistry::getRepository(
+                CartCarrierDropOffMapping::getClassName()
+            );
+
+            $query = new QueryFilter();
+            $query->where('cartId', '=', (string)$shopOrder->id_cart)
+                ->where('carrierReferenceId', '=', (string)$shopOrder->id_carrier);
+
+            /** @var \Packlink\PrestaShop\Classes\Entities\CartCarrierDropOffMapping $mapping */
+            $mapping = $repository->selectOne($query);
+
+            if ($mapping) {
+                $dropOff = $mapping->getDropOff();
+
+                return (string)$dropOff['id'];
+            }
+        } catch (\Exception $e) {
+            return null;
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns packlink address from shop address.
+     *
+     * @param PrestaShopOrder $shopOrder
+     *
+     * @return \Packlink\BusinessLogic\Order\Objects\Address
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function getAddress(PrestaShopOrder $shopOrder)
+    {
+        $deliveryAddressId = (int)$shopOrder->id_address_delivery;
+
+        $shippingAddress = new Address();
+        $deliveryAddress = new PrestaShopAddress($deliveryAddressId);
+        $customer = new \Customer($shopOrder->id_customer);
+        $country = new \Country($deliveryAddress->id_country);
+
+        if ($country !== null) {
```

Comment:
In this case $country is not always !== null?

Reply:
Updated to correctly validate the loaded object.